File: src/java/org/javaturk/dp/ch02/factory/solution/Director.java
package org.javaturk.dp.ch02.factory.solution;


public class Director extends Manager {
	protected double bonus;
	
	public Director(int no, String name, int year, String workingDepartment, String managingDepartment, double bonus) {
		super(no, name, year, workingDepartment, managingDepartment);
		this.bonus = bonus;
	}
	
	public void work(){
		System.out.println("Director is working!");
		manage();
	}
	
	public void manage(){
		System.out.println("Director manages whole company!");
		makeAStrategicPlan();
	}
	
	public void makeAStrategicPlan(){
		System.out.println("Director makes a strategic plan for the company!");
	}
	
	public double calculateSalary(){
		return super.calculateSalary() + bonus;
	}
	
	public void printInfo(){
		System.out.println("\nDirector Info");
		super.printInfo();
	}

	@Override
	public String toString() {
		return "Director [bonus=" + bonus + ", departmentManaged=" + departmentManaged + ", no=" + no + ", name="
				+ name + ", year=" + year + ", department=" + department + "]";
	}
}

File: src/typescript/com/gokselkucuksahin/dp/creational-patterns/factory/employee1/director.ts
import {Manager} from "./manager";

export class Director extends Manager {
  protected bonus: number;

  constructor(id: number, name: string, year: number,
              workingDepartment: string, managingDepartment: string, bonus: number) {
    super(id, name, year, workingDepartment, managingDepartment);
    this.bonus = bonus;
  }

  work() {
    console.log("Director is working!");
    this.manage();
  }

  manage() {
    console.log("Director manages whole company!");
    this.makeAStrategicPlan();
  }

  makeAStrategicPlan(): void {
    console.log("Director makes a strategic plan for the company!");
  }

  calculateSalary(): number {
    return super.calculateSalary() + this.bonus;
  }

  getBonus(): number {
    return this.bonus;
  }

  printInfo() {
    console.log("\nDirector Info");
    super.printInfo();
  }

  toString(): string {
    return `Director [bonus=${this.bonus}, departmentManaged=${this.departmentManaged}, id=${this.id}, name=${this.name}, year=${this.year}, department=${this.department}, salary=${this.salary}]`;
  }
}